### PR TITLE
Create BytesMut for Json with initial capacity

### PR DESCRIPTION
**Pull request by ryanfowler**
This PR initializes the `BytesMut` for serializing `Json` with a small initial capacity to reduce allocations as the inner `Vec<u8>` grows.

A [recent PR](https://github.com/tokio-rs/axum/pull/1178) removed an unnecessary copy by introducing `BytesMut` and switching away from `serde_json::to_vec`, but I noticed a small regression that I believe is due to `serde_json::to_vec` creating a `Vec<u8>` with an initial capacity ([code](https://github.com/serde-rs/json/blob/v1.0.82/src/ser.rs#L2189)). I figured matching the initial capacity of "128 bytes" made sense. In a simple benchmark like below (mostly pulled from [this comment](https://github.com/tokio-rs/axum/issues/1177#issuecomment-1194007420)), I'm seeing a modest ~1% speed bump.

```rust
use axum::{routing::get, Json, Router};

#[tokio::main]
async fn main() {
    let app = Router::new().route("/json", get(json));

    axum::Server::bind(&"0.0.0.0:8000".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}

async fn json() -> Json<Message> {
    let message = Message {
        name: "http-benchmark",
        version: "v0.1.0",
        message: "Hello, World!",
    };
    Json(message)
}

#[derive(serde::Serialize)]
pub struct Message {
    pub name: &'static str,
    pub version: &'static str,
    pub message: &'static str,
}
```

### CODE REVIEW

File: axum/src/json.rs [hunk 1]
```diff
@@ -186,7 +186,7 @@ where
     T: Serialize,
 {
     fn into_response(self) -> Response {
-        let mut buf = BytesMut::new().writer();
+        let mut buf = BytesMut::with_capacity(128).writer();
```
Comment (davidpdrsn) [inline axum/src/json.rs:189]: Maybe add a comment about why `128` specifically. Can just link to the code in `serde_json`.

**Review by davidpdrsn — COMMENTED** [review]
Makes sense!

[actix-web uses an initial capacity of 8192 bytes](https://docs.rs/actix-web/4.1.0/src/actix_web/types/json.rs.html#371). I'm wondering if we should benchmark a slightly bigger initial capacity as well 🤔

File: axum/src/json.rs (hunk 1 above)
Comment (jplatte) [reply axum/src/json.rs:191]: Could also use a docs.rs link: https://docs.rs/serde_json/1.0.82/src/serde_json/ser.rs.html#2189
(could be longer-lived, but who knows)

**Comment by ryanfowler** [comment]
> [actix-web uses an initial capacity of 8192 bytes](https://docs.rs/actix-web/4.1.0/src/actix_web/types/json.rs.html#371). I'm wondering if we should benchmark a slightly bigger initial capacity as well 🤔

Oh, interesting! I suppose it completely depends on the size of JSON being serialized. I would guess that JSON responses in real-world servers are usually >128 bytes...

File: axum/src/json.rs (hunk 1 above)
Comment (ryanfowler) [reply axum/src/json.rs:191]: Done!

**Review by jplatte — APPROVED** [review]
Intuitively, 128 seems like a good initial capacity to me but 8k seems a little much.

**Review by davidpdrsn — APPROVED** [review]
Yep its cool with me as well. We can always tweak later.

**Comment by ryanfowler** [comment]
Hmm it appears that the required checks for the `main` branch may need to be updated for [this](https://github.com/tokio-rs/axum/commit/3bcc8fb07d0abfb8548c701b7028a9f63b1cab9b) naming change made to CI

**Comment by davidpdrsn** [comment]
Yeah I'll do that separately.

Thanks!

**Pull request merged**